Add a page for composefs (#663)
coreos/fedora-coreos-docs: modules/ROOT/pages/composefs.adoc

= ComposeFS

Fedora CoreOS introduced ComposeFS enabled by default starting in Fedora 41. ComposeFS is an overlay filesystem where the data comes from the usual ostree deployement, and

> metadata are in the composeFS file. The result is a truely read-only root (`/`) filesystem, increasing the system integrity and robustness,

Reviewer: Upstream mostly uses `composefs` (https://github.com/search?q=repo%3Acontainers%2Fcomposefs%20composefs&type=code) only uses `Composefs` at the beginning of sentences, so let's use `composefs` everywhere.

This is a first step towards a full verification of filesystem integrity, even at runtime.

== What does it change ? 

The main visible change will be that the root filesystem (/) is now small and full (a few MB, 100% used).

> The real root is mounted in /sysroot and most of the data is stored in /var. 

Author: copy/paste leftover from the change proposal :D

> == How to disable composeFS

Author: copy/paste leftover from the change proposal :D

ComposeFS can be disabled through a kernel argument: `ostree.prepare-root.composefs=0`.

.Disabling composeFS mount
[source,yaml,subs="attributes"]
----
variant: fcos
version: {butane-latest-stable-spec}
kernel_arguments:
  should_exist:
    - ostree.prepare-root.composefs=0
----

== Known issues

=== Kdump

> Right now, this prevents kdump from writing the memory dumps as it get confused by the read-only filesystem.

Reviewer: Isn't the issue during the generation of the initramfs?

Author: ah yes, good point. But it's worth noting that setting a remote target (ssh, nfs) is a workaround

If you want to use kdump and export kernels dumps to the local machine, composeFS must be disabled.

> The kdump upstream developpers are working on a fix. We will update this page when the workaround is no longer needed.

Reviewer: Let's link to the issue on the kdump tracker.

=== Top-level directories

Another consequence is that it is now impossible to create top-level direcories in `/`. Those are usually mount points.
There are a few ways around this: 
 - Disable composeFS as showed above.
 - Enable transient root by adding `root.transient=true` in `/etc/ostree/prepare-root.conf`. Please note that all changes created
(even in `/etc`) will be lost on upgrades. More details in https://ostreedev.github.io/ostree/man/ostree-prepare-root.html[ostree documentation].

>  - Derive a container and create your required top-level directories there. 

Author: Ywah i wasn't too sure. Thanks for the confirmation

= Composefs

Fedora CoreOS introduced composefs enabled by default starting in Fedora 41. Composefs is an overlay filesystem where the data comes from the usual ostree deployement, and
metadata are in the composefs file. The result is a truely read-only root (`/`) filesystem, increasing the system integrity and robustness,

This is a first step towards a full verification of filesystem integrity, even at runtime.

== What does it change ? 

The main visible change will be that the root filesystem (/) is now small and full (a few MB, 100% used).

== Known issues

=== Kdump

Right now, this prevents kdump from generating it's initramfs as it get confused by the read-only filesystem.
If you want to use kdump and export kernels dumps to the local machine, composefs must be disabled.
A workaround is to configure kdump with a remote target such as ssh or nfs.
The kdump upstream developpers are working on a fix. We will update this page when the workaround is no longer needed.

> https://github.com/rhkdump/kdump-utils/pull/28

Reviewer: ```suggestion
See [https://github.com/rhkdump/kdump-utils/pull/28 rhkdump/kdump-utils#28].
```

=== Top-level directories

> Another consequence is that it is now impossible to create top-level direcories in `/`. Those are usually mount points.

Reviewer: ```suggestion
Another consequence is that it is now impossible to dynamically create top-level directories in `/`. A common use case for those top level directories is to use them as mount points. We recommend using sub directories in `/var` instead.
```

Author: `dynamically` suggest that there is a way around IMO.

> Currently, the only way around that is to disable composefs as showed above.

Reviewer: ```suggestion
Currently, the only way around that is to disable composefs as showed below.
```

Author: good catch :)

= Composefs

Fedora CoreOS introduced composefs enabled by default starting in Fedora 41. Composefs is an overlay filesystem where the data comes from the usual ostree deployement, and
metadata are in the composefs file. The result is a truely read-only root (`/`) filesystem, increasing the system integrity and robustness,

This is a first step towards a full verification of filesystem integrity, even at runtime.

== What does it change ? 

The main visible change will be that the root filesystem (/) is now small and full (a few MB, 100% used).

== Known issues

> === Kdump

Reviewer 2: i'm really hoping we can have kdump fixed before F41 gets promoted to our `stable` stream.

> Right now, this prevents kdump from generating it's initramfs as it get confused by the read-only filesystem.

Reviewer 2: ```suggestion
Right now, this prevents kdump from generating its initramfs as it gets confused by the read-only filesystem.
```

If you want to use kdump and export kernels dumps to the local machine, composefs must be disabled.
A workaround is to configure kdump with a remote target such as ssh or nfs.
The kdump upstream developers are working on a fix. We will update this page when the workaround is no longer needed.

> See [https://github.com/rhkdump/kdump-utils/pull/28 rhkdump/kdump-utils#28].

Reviewer 2: since this is adoc:

```suggestion
See https://github.com/rhkdump/kdump-utils/pull/28[rhkdump/kdump-utils#28].
```
